core/storage/trie/polkadot_trie/polkadot_trie_impl.cpp

```diff
@@ -276,176 +284,181 @@ namespace kagome::storage::trie {
     return root_ == nullptr;
   }
 
+  outcome::result<void> deleteNode(
+      PolkadotTrie::NodePtr &,
+      const KeyNibbles &,
+      const PolkadotTrieImpl::ChildRetrieveFunctor &retrieveChild);
+
   outcome::result<void> PolkadotTrieImpl::remove(const common::Buffer &key) {
-    if (root_) {
-      auto key_nibbles = PolkadotCodec::keyToNibbles(key);
-      // delete node will fetch nodes that it needs from the storage (the nodes
-      // typically are a path in the trie) and work on them in memory
-      OUTCOME_TRY(n, deleteNode(root_, key_nibbles));
-      // afterwards, the nodes are written back to the storage and the new trie
-      // root hash is obtained
-      root_ = n;
-    }
-    return outcome::success();
+    auto key_nibbles = PolkadotCodec::keyToNibbles(key);
+    // delete node will fetch nodes that it needs from the storage (the
+    // nodes typically are a path in the trie) and work on them in memory
+    return deleteNode(root_, key_nibbles, retrieve_child_);
   }
 
-  outcome::result<PolkadotTrie::NodePtr> PolkadotTrieImpl::deleteNode(
-      NodePtr parent, const KeyNibbles &key_nibbles) {
+  outcome::result<void> handleDeletion(
+      PolkadotTrie::NodePtr &, const PolkadotTrieImpl::ChildRetrieveFunctor &);
+
+  outcome::result<void> deleteNode(
+      PolkadotTrie::NodePtr &parent,
+      const KeyNibbles &key,
+      const PolkadotTrieImpl::ChildRetrieveFunctor &retrieveChild) {
     if (parent == nullptr) {
-      return nullptr;
+      return outcome::success();
     }
-    using T = PolkadotNode::Type;
-    auto newRoot = parent;
-    switch (parent->getTrieType()) {
-      case T::BranchWithValue:
-      case T::BranchEmptyValue: {
-        auto length = getCommonPrefixLength(parent->key_nibbles, key_nibbles);
-        auto parent_as_branch = std::dynamic_pointer_cast<BranchNode>(parent);
-        if (parent->key_nibbles == key_nibbles or key_nibbles.empty()) {
-          parent->value = boost::none;
-          newRoot = parent;
-        } else {
-          OUTCOME_TRY(child,
-                      retrieveChild(parent_as_branch, key_nibbles[length]));
-          OUTCOME_TRY(n, deleteNode(child, key_nibbles.subspan(length + 1)));
-          newRoot = parent;
-          parent_as_branch->children.at(key_nibbles[length]) = n;
-        }
-        OUTCOME_TRY(n, handleDeletion(parent_as_branch, newRoot, key_nibbles));
-        return std::move(n);
+    if (parent->isBranch()) {
+      if (parent->key_nibbles == key) {
+        parent->value = boost::none;
+      } else {
+        auto length = getCommonPrefixLength(parent->key_nibbles, key);
+        OUTCOME_TRY(ch,
+                    retrieveChild(std::dynamic_pointer_cast<BranchNode>(parent),
+                                  key[length]));
+        auto &child =
+            dynamic_cast<BranchNode &>(*parent.get()).children.at(key[length]);
```

Comment:
Usage of `dynamic_cast` for no reason. You don't check the value for nullptr before usage anyway.
```suggestion
        BranchNode &child = (*parent.get()).children.at(key[length]);
```

Reply:
I exactly know it is a BranchNode, there was a condition before. Check for nullptr was earlier. Shared_ptrs, node polymorphism are redundant here, but I can't change it in this little PR.

Reply:
Parent is of base type NodePtr.